src/providers/gif.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{
+    render::{display::ContentProvider, scheduler::ContentWrapper, gif},
+    scheduler::CONTENT_PROVIDERS,
+};
+use anyhow::Result;
+use apex_hardware::FrameBuffer;
+use async_stream::try_stream;
+use config::Config;
+use embedded_graphics::{
+    geometry::Point,
+};
+use futures::Stream;
+use linkme::distributed_slice;
+use log::info;
+use tokio::{
+    time,
+    time::{Duration, MissedTickBehavior},
+};
+use std::fs::File;
+
+
+
+
+#[doc(hidden)]
+#[distributed_slice(CONTENT_PROVIDERS)]
+pub static PROVIDER_INIT: fn(&Config) -> Result<Box<dyn ContentWrapper>> = register_callback;
+
+#[doc(hidden)]
+#[allow(clippy::unnecessary_wraps)]
+fn register_callback(config: &Config) -> Result<Box<dyn ContentWrapper>> {
+    info!("Registering Gif display source.");
+
+    let gif_path = config.get_str("gif.path").unwrap();
+    let gif_file = File::open(&gif_path);
+
+    let gif = match gif_file {
+        Ok(file) => gif::Gif::new(Point::new(0, 0), Point::new(128, 40), file),
+        Err(err) => {
+            log::error!("Failed to open GIF file '{}': {}", gif_path, err);
+			
+            // Use the `new_error` function to create an error GIF
+            let error_gif = gif::Gif::new_error(Point::new(0, 0), Point::new(128, 40));
+            error_gif
+        }
+    };
+
+    Ok(Box::new(Gif {gif}))
+}
+
+pub struct Gif {
+	gif: gif::Gif
+}
+
+impl Gif {
+    pub fn render(&self) -> Result<FrameBuffer> {
+        let mut buffer = FrameBuffer::new();
+
+		self.gif.draw(&mut buffer);
+
+        Ok(buffer)
+    }
+}
+
+impl ContentProvider for Gif {
+    type ContentStream<'a> = impl Stream<Item = Result<FrameBuffer>> + 'a;
+
+    // This needs to be enabled until full GAT support is here
+    #[allow(clippy::needless_lifetimes)]
+    fn stream<'this>(&'this mut self) -> Result<Self::ContentStream<'this>> {
+        let mut interval = time::interval(Duration::from_millis(10)); 
+		//the delays in gifs are in increments of 10 ms
+		//https://docs.rs/gif/latest/gif/struct.Frame.html#structfield.delay

```

Comment:
That's weird, most of the gifs I tested had a delay field that wasn't null (I wish I had the right to share them, but sadly, I don't).

I'm actually using the delay in the renderer (to calculate if it's actually time to change the frame), if no delay is found in the gif, it will run 1 frame per 10 ms.

Moreover, I quite sure the delay field is used in modern browser (/gallery), as otherwise all gifs would play at 100fps.

So I'm not sure if we should implement a config toggle, if we need, then I will also need to add the feature in the renderer.

Reply:
Fair enough, your explanation makes sense.